### DISCUSSION TITLE
Weird crash when creating new images (using wrong constructor?)

### DISCUSSION THREAD

**Discussion by tsftd** [ImageMagick/ImageMagick#2499] [Help] [answered]
I'm not sure at what point it happened, but at some point over the past few months a previously-working project has broken. Imagemagick was updated recently, but I don't remember if it was before or after or when the project broke.

Whenever I try to construct an image using Geometry,Color it will crash with the error `Magick: unable to open image 'xc:#FFFFFFFFFFFF': No such file or directory @ error/blob.c/OpenBlob/3496` (where #FFFFFFFFFFFF is the color). From what I can tell, it's interpreting the Color argument as a filename for some reason.

Exact code: `dummyimage=Image(Magick::Geometry("1024x1024"),Magick::Color("white"));`

Note that the same issue persists whether I input a color name or '#FFFFFFFFFFFF' directly, and whether I construct the Geometry as "1024x1024" or 1024,1024 , and whether I create const Magick::Geometry and const Magick::Color beforehand and use them as variables in the constructor.

I've also confirmed that the current user has ownership and full control over the entire tree containing all of my programming stuff (including the Magick++ dependencies), as apparently the 'No such file or directory' error is usually related to that. In my case, of course, it should not make a difference since I'm not trying to open a file in the first place.

This section of the code has not changed in years.

MSVC v15.9.23, Imagemagick v7.0.10-Q16-HDRI

**Comment by snibgo** [comment]
What happens when you run the equivalent magick command:
```
magick -size 1024x1024 xc:white NULL:
```
What is the text output from "magick -version"?

After you installed a new version of IM, I assume you re-built your program(s)?

Do the supplied test programs build and run successfully?

What is the text output from your program when magick debug is used?

EDIT to add: "No such file or directory" is a generic Windows error message when it has nothing else to report. It doesn't necessarily mean the program was trying to access a file.

**Reply by tsftd** [reply]
Oh, that's good information, it's probably an issue with the install since cmdline gives:
`C:\WINDOWS\system32>magick -size 1024x1024 xc:white NULL:
magick: unable to open image 'xc:white': No such file or directory @ error/blob.c/OpenBlob/3496.
magick: no decode delegate for this image format `' @ error/constitute.c/ReadImage/562.`

However, reinstalling does not fix the issue...

`C:\WINDOWS\system32>magick -version
Version: ImageMagick 7.0.10-16 Q16 x64 2020-05-31 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 180040629
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib`

What's strange, is that despite listing the jpeg delegate as built-in, this happens:

`C:\SyncThing\Laptop\Programs\Programming\ImageMagick-7.0.10-Q16-HDRI>magick convert melonbear.jpg test.png
convert: no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/562.
convert: no images defined `test.png' @ error/convert.c/ConvertImageCommand/3282.`

Yes, I am running in debug mode from the IDE -- it is rebuilding the project on each change.

Unfortunately, the test programs throw a million errors, such as asking for files that don't exist:
`1>LINK : fatal error LNK1181: cannot open input file 'CORE_RL_magick_.lib'`

Hopefully we can figure out what's causing the issue without me having to get the test programs working, as I remember last time I had to do that it was a nightmare.

Sorry, I'm not sure what you mean by 'magick debug' -- do you mean compiling debug dlls?

P.S. I'm on Japan time, so sorry for the lag in responses.

**Comment by snibgo** [comment, marked as answer]
For debug, you can call...
```
SetLogEventMask("all");
```
... early in your program. Or set an environment variable. This "debug" is an internal IM feature, not MSVC debug mode. I'm told the MSVC debug mode upsets IM, so it shouldn't be used. I don't use MSVC, so can't advise.

However, as "magick" doesn't work, don't bother about debug.

The failure of...
```
magick -size 1024x1024 xc:white NULL:
```
... means the installation is bad. How did you install? I hope you went to https://www.imagemagick.org/script/download.php#windows and downloaded one of the .exe files (which one?) and ran it, and followed a dialogue to install ImageMagick. If you installed by some other method, eg by copying files, then that is probably the problem.

Perhaps you have multiple versions installed, and are using some components from one installation and other components from a different installation. It is possible to have multiple versions installed (I have about 25 versions), but they must be kept separate, and in different directories.

Regardless, I suggest you uninstall all versions of ImageMagick. In Windows 8.1, goto Control Panel, Programs and features. For every ImageMagick on the list, right-click and click "Uninstall". 

Then goto whatever directory you installed ImageMagick in, and check it has no IM files.

Then goto https://www.imagemagick.org/script/download.php#windows (or use the installation .exe you prevously downloaded) and install as above. Then verify IM is working, eg try:
```
magick logo: logo.gif
magick identify logo.gif
magick -size 1024x1024 xc:white NULL:
```

If they all work, then re-build your program, and try running it.

**Reply by tsftd** [reply]
Yes, IIRC there are issues with debug mode, I just build release mode and debug that manually. Not exactly good form, I know, but it works.

I had actually uninstalled & reinstalled before my last post. I did get the official file from the website -- ImageMagick-7.0.10-16-Q16-HDRI-x64-dll.exe in this case (was downloaded last June).

My _guess_ is that I didn't uninstall the previous IM properly (though it doesn't show up in my installed programs list) when I upgraded, but I'm afraid that it'll be difficult-to-impossible to track down exactly what the problem is and get it resolved (since simply uninstalling and reinstalling the current version doesn't appear to stomp on it). I do know that I installed it to a different directory at least, since it has the version number in it.

The good news is that I was planning on reinstalling windows anyway, as it's become pretty sketchy (last reinstall was I think about 2.5 years ago and I usually reinstall about once a year). I was actually backing everything up to my desktop when this issue cropped up but I wanted to ensure that it was working before nuking the system.

I'll try and get it taken care of this weekend and see if that resolves the problem. Stuff like this is why I rarely ever upgrade dependencies (last time was I think 3-4 years ago), but to be fair I'm not a professional programmer so it's as much down to the janky way I do things.

I'll consider this resolved, but post back confirmation assuming windows reinstall fixes it just in case someone else has a similar issue in the future.
